### PR TITLE
Add weak heuristics in our designer scanning code to improve experience during things like configuration switch.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Current approach leads to projects changing their designability, which can break the designer, even though it actually works in these scenarios.  By adding a couple of weak heuristics (that we discussed with WinForms designer team) we can be much more robust.

We will likely backport this to 17.10 as well.  But getting review in Main first.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
@Cosifne @genlu ptal.

**Comment by CyrusNajmabadi** [comment]
@genlu @Cosifne ptal

File: src/Features/Core/Portable/DesignerAttribute/DesignerAttributeDiscoveryService.cs [hunk 1]
```diff
@@ -305,6 +324,33 @@ private static async ValueTask<bool> HasDesignerCategoryTypeAsync(Project projec
...
+                if (current is IErrorTypeSymbol errorType)
+                    current = TryMapToNonErrorType(compilation, errorType);
+            }
+        }
+
+        static INamedTypeSymbol? TryMapToNonErrorType(Compilation compilation, IErrorTypeSymbol errorType)
+        {
+            foreach (var wellKnownNamespace in s_wellKnownDesignerNamespaces)
+            {
+                var wellKnownType = compilation.GetTypeByMetadataName($"{wellKnownNamespace}.{errorType.Name}");
```
Comment (Cosifne) [inline src/Features/Core/Portable/DesignerAttribute/DesignerAttributeDiscoveryService.cs:344]: 🤔When will this happen to user? Namespace misspelling, why the namespace in SemanticModel is wrong?

**Comment by Cosifne** [comment]
Approach is Ok, but I am curious about what's the affected secnario?

File: src/Features/Core/Portable/DesignerAttribute/DesignerAttributeDiscoveryService.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/DesignerAttribute/DesignerAttributeDiscoveryService.cs:344]: Basically, as a restore happens, we can lose and then get references again.   Also, generators may not run again until we know about them again, so we lost global usings. 

This keeps everything still looking good (and usable for the designer side) while we get back into a good state again.

**Review by Cosifne — APPROVED** [review]

**Pull request merged**